InsiteTeamTask.Services/Attendence/AttendenceService.cs

```diff
@@ -0,0 +1,53 @@
+﻿using InsiteTeamTask.Data.DTOs;
+using InsiteTeamTask.Data.Repositories.Interfaces;
+using InsiteTeamTask.Logic.Attendence.Actions;
+
+namespace InsiteTeamTask.Logic.Attendence
+{
+    class AttendenceService : IAttendenceService
+    {
+        private readonly IProductRepository _productRepository;
+        private readonly IAttendenceActionFactory _attendenceActionFactory;
+
+        public AttendenceService(IProductRepository productRepository, IAttendenceActionFactory attendenceActionFactory)
+        {
+            _productRepository = productRepository;
+            _attendenceActionFactory = attendenceActionFactory;
+        }
+
+        public AttendenceDto GetAttendenceForGameNumberAndSeasonNumber(int gameNumber, int seasonNumber)
+        {
+            var attendenceDto = new AttendenceDto();
+            var productDtos = _productRepository.GetProductByGameNumberAndSeasonNumber(gameNumber, seasonNumber);
+
+            //While this can be done in a single linq expression, that would close it for modification. Also make it harder to read.
+            foreach (var productDto in productDtos)
+            {
+                attendenceDto = GetAttendenceFromProductDto(productDto, attendenceDto);
+            }
+
+            return attendenceDto;
+        }
+
+        public AttendenceDto GetAttendanceForProductCode(string productCode)
+        {
+            var attendenceDto = new AttendenceDto();
+            var productDto = _productRepository.GetProductByCode(productCode);
+
+            if (productDto != null)
+            {
+                attendenceDto = GetAttendenceFromProductDto(productDto, attendenceDto);
+            }
+
+            return attendenceDto;
+        }
+
+        private AttendenceDto GetAttendenceFromProductDto(ProductDto productDto, AttendenceDto attendenceDto)
+        {
+            //Instead of using a switch statement use a dictionary of actions that can easily be expanded
+            attendenceDto = _attendenceActionFactory.GetAction(productDto.Type).DoAction(productDto, attendenceDto);
```

Comment:
Okay, time to tackle your structure.

When a user makes a request to get attendance by a game and season, the following steps happen:
- The `AttendanceController` calls an `IAttendanceViewService`
- The `IAttendanceViewService` calls an `IAttendanceService`
- The `IAttendanceService` calls an `IProductRepository`
- The `IProductRepository` retrieves relevant `Product` objects from an `IDataService`, and returns them as DTOs
- The `IAttendanceService` calls an `IAttendanceFactory` for each returned `Product` which returns an `IAttendanceAction` depending on that `Product`'s Type
- The `IAttendanceAction` calls either an `ITicketRepository` or an `IMemberRepository` with the `ProductId` passed through, and adds the returned objects to the `AttendanceDto`

This is probably about 4 steps too many, and a couple too many levels of abstraction to match a few object types together based on a GameId and SeasonId. This code is a nightmare to follow for anyone who didn't write it or hasn't spent at least an hour looking at it, and you definitely couldn't explain this to someone offhandedly if they needed to work on it.

To simplify your structure:
- Controller calls Service
- Service fetches data (Tickets, and Members) from their respective repositories
- Service filters Tickets and Members by gameId and seasonId
- Service converts these collections to `Attendance` collections
- Service returns a single `Attendance` collection

It's good to see that you know Factory and Command patterns, but I'm second-guessing if you know how and where to adeptly use them